### ISSUE TITLE
Python analysis code rework proposal

### ISSUE THREAD

**Issue by slundqui** [PetaVision/OpenPV#34]
Hi all,

TL;DR Octave peg in Python hole makes writepvpfile difficult. If you're not using Python analysis tools, this issue is not important to you.

I'm currently looking over our Python analysis tools (thanks to @maxtheiler for putting it together). I would like to propose some changes to the code. Currently, readpvpfile returns an PVObject, which itself contains a header field. Furthermore, through some Python hack that I'm not too familiar with, we are able to index into this object for a frame, followed by a dot notation into the time scalar and values matrix, i.e.

``` python
#Get header
pvObj.header
#Access value
pvObj[frameIdx].values[yIdx, xIdx, fIdx]
```

This was done to keep the Python and Octave interfaces similar to each other.

The problem this architecture raises is when we would like to create a custom pvp file via a "writepvpfile" function. The current implementation requires you to build the pvObject by hand, including filling in all of the header fields (@maxtheiler please correct me if I'm wrong on this matter). In contrast, writepvpfile in Octave only requires the data, with most of the header fields being inferred through the data structure (an exception is the case of sparse values, where we must explicitly tell the function the size of the matrix). The Octave architecture was advantageous in 2 ways: (1) it was easy to read a pvp file, change something simple in the data, and feed the same structure with the changed data right back to writepvpfile; (2) it was easy to create the data structure from scratch, as writepvpfile expected a parameter that used native Octave's built-in data structures, i.e. a cell array with the fields values and time, which was a matrix and scalar respectively. In contrast, the Python writepvpfile only supports (1).

I see several solutions to this problem: 
1. We write a pvObject constructor that takes a native Python/Numpy data structure and generates the header field of the pvObject from that.
2. We separate the header from the pvObject, where readpvpfile returns a tuple (one for the data, one for the header, analogous to what Octave's readpvpfile is returning), i.e.

``` python
(dataObj, header) = readpvpfile(pvpFilename)
```

However, we still would need a constructor that allows us to create a pvObject from native Python/Numpy data structures.
3. We boot the idea of a pvObject all together, and instead focus on using Python/Numpy data structures as is (while keeping the header separate). This would bring about several differences between the Octave and Python versions of these analysis functions (but personally, I think this is okay, as trying to force an Octave peg into a Python hole doesn't make too much sense to me).

I personally like option 3. Off the top of my head, a list of Python dictionaries with the fields "time" and "values", where data would be a Numpy array as it is now would do the trick. Accessing the data would be as follows: 

``` python
#Set time
data[frameIdx]["time"] = 1
#Allocate Numpy array for dense values
data[frameIdx]["values"] = np.ones((ny, nx, nf))
#Access data value
data[frameIdx]["values"][yIdx, xIdx, fIdx]
```

Another option would be have an outer data dictionary, with a list of Numpy arrays: 

``` python
#Set time
data["time"][frameIdx] = 1
#Allocate Numpy array for dense values
data["values"][frameIdx] = np.ones((ny, nx, nf))
#Access data value
data["values"][frameIdx][yIdx, xIdx, fIdx]
```

As I'm sure there's memory overhead in Python dictionaries, I like the latter data structure better.

Sorry for the lengthy post, as I'm mostly writing this down to get the idea laid out on paper, as well as getting an idea of who is actually using the Python analysis scripts. I'm under the assumption that @dpaiton and I are the only ones currently using the Python analysis tools, so the damage of overhauling the Python analysis scripts would be limited.

Let me know your thoughts,
Sheng

**Comment by maxtheiler** [comment]
Hey guys,
This is also a response to [Issue #33](https://github.com/PetaVision/OpenPV/issues/33), just weighing in on the pvtools as they now stand.

**Point 1**
TLDR: I think the objects are pretty close to optimal lookup speed as written, plus we can have python-looking and octave-looking call syntax _both_ be valid syntaxes if we use homemade objects like PV_Object and DataFrame.

The current PV_Object is basically a _list_ of _tuples_. It's actually a "PV_Object" of "DataFrames", but PV_Object inherits list and DataFrame inherits tuple. The 'PV_Object' is just a list with some attached extra dressing, i.e. a header attribute and some generic analysis functions. The 'DataFrame' tuple has been reworked to understand a call to its '.time' attribute as its [0]th index and a call to its '.values' attribute to its [1]th index. The '.values'/[1] index already points to a numpy array that can be spliced and diced efficiently in all the usual numpy ways.

As Sheng and Dylan have mentioned, the original point of these modifications was to fit a python peg in an octave-shaped hole. The revised calling syntax uses an 'attribute' of the object, which looks like "anObject.anAttrbute" rather than the "AnObject['attribute_string_index']" syntax used by a vanilla python dict object, purely because it feels more like octave styntax for the person writing the code. But conventional tuple indexing can still be used with a DataFrame object as written, so

``` python
pvObj[frameIdx].time, pvObj[frameIdx].values[yIdx, xIdx, fIdx]
```

is the same call as

``` python
pvObj[frameIdx][0], pvObj[frameIdx][1][yIdx, xIdx, fIdx]
```

I believe (though I might be wrong about this) the cost in terms of lookup times for using these homemade objects rather than native python objects is not a problem. From what I understand object attribute lookups are accomplished as dict lookups, so "thisFrame.values[x,y,f]" and "thisFrame['values'][x,y,f]" are basically just two ways of framing the same underlying algorithm, with the former having the advantage of looking octave-y for the user.

The main point I'm trying to make is that I think by using homemade objects we can have the best of both worlds. DataFrame might be better off as an inheritor of 'dict', rather than of 'tuple' as it is now. But using a homemade object would allow the user to write _either_ "pvObj[frameIdx].values[yIdx, xIdx, fIdx]" _or_ "pvObj[frameIdx]['values'][yIdx, xIdx, fIdx]" interchangeably, without (I think) any significant differences in lookup times. Meanwhile the lookups for individual values are all dome with numpy, so the optimization there is probably as good as it's going to get while keeping everything grouped as a single object.

**Point 2:**
TLDR: DataFrame should probably extend numpy.array, rather than tuple

If we're happy to jettison the old way, an even better option for some hypothetical pure-python user would probably be to set up DataFrame as an inheritor of numpy array itself, rather than as a dict or tuple holding a numpy array. This would save a dict/tuple lookup and allow the sleek calling syntax:

``` python
pvObj[frameIdx][yIdx, xIdx, fIdx]
```

".time" could remain an attribute of the (np.array-inheriting) DataFrame class, allowing:

``` python
pvObj[frameIdx].time
```

It's probably also possible to have a redundant attribute that just points to the object itself, which would allow the octave-y "pvObj[frameIdx].values[yIdx, xIdx, fIdx]" syntax to coexist with the above sleekness. Currently, calling the DataFrame object via pvObj[frameIdx] returns the whole object, so if you're working in interactive mode you get both the time entry and the values entry from a single call. This is a pretty dubious advantage of using a tuple/dict rather than a custom numpy array with a time attribute, but losing that functionality the only drawback to this method I can think of at the moment.

**Point 3:**
TLDR: writepvpfile.py remains half-baked.

As Sheng points out, writepvpfile.py is pretty lame at the moment because it requires you to input a PV_Object, which is technically a homemade object (but can be thought of as a list with some extras), and accepts no substitutes. Different pvpfile types are written differently as binary files (some repeat the header for each frame, some are sparse, etc.), but it would not be hard to get writepvpfile.py to understand inputs formatted as, say, 'a list of numpy arrays' and to cast them appropriately before writing. Inferring complete headers out of 'from-scratch' values inputs might take a bit more thought.

**Comment by dpaiton** [comment]
Thank you @maxtheiler for making this more clear and @slundqui for putting the time into writing out exactly what should be addressed. I do have a question, though. My primary concern is not really lookup times, but performing matrix operations. For example, if I want to load in a PVP file of activities perform some large matrix operation on them, I would need to convert the object into a proper numpy matrix (sparse or otherwise). I do like the benefits of having the custom object, but I believe the data should be formed into a [frame, fidx, yidx, xidx] tensor. If I am not mistaken, this is not the case currently. Right?

**Comment by maxtheiler** [comment]
You're correct. Incorporating frames as part of a 4d tensor would mean faster matrix operations on the entire object, and that is not how the framework is set up currently.

**Comment by slundqui** [comment]
The reason why the frame dimension is separate is due to the fact that not all values from frame to frame stay in the same shape. Specifically, sparse matrices in the index/value format had a variable length going from frame to frame. One option here is to use Numpy's sparse matrix data structure instead. However, the downside is that it only supports 2D matrices, as opposed to a 4D tensor, so any access to the array must follow an index conversion.

For writepvpfile, I guess my point is that the advantage of having a pvobject wrapper is the syntactic sugar of the dot operator, with the tradeoff that we introduce a complex custom data structure. Personally, I don't think the advantages outweigh the tradeoff. If I wanted to write a custom pvp file where I need to build some sort of Python/Numpy native data structure anyways, why not just use that data structure as the format of readpvpfile?

**Comment by dpaiton** [comment]
I forgot about the fact that different images will have different sparsity levels. Per image, I have converted the PVObject to a sparse matrix by vectorizing the feature, y, and x dimensions. As long as you know the index cycling order you can undo this. One advantage to saving the data as a scipy sparse matrix is that it has a small memory footprint and allows for increased efficiency with matrix operations (as opposed to a dense matrix). However, those matrix operations may not be intuitive or even possible with the data vectorized. And if one has to "densify" the sparse matrix into an ndarray then we are not any more efficient than the current implementation.

**Comment by slundqui** [comment]
I'm okay with using sparse matrices via vectorizing the y, x, and feature dimensions, as this new data structure will allow us to combine the frames dimension in with the data dimension. As the Octave code returns a similar sparse data structure, the disadvantages of having to densify the sparse structure were already there, so representing the data with a Numpy sparse matrix seems like a good idea. With this in mind, can we make a decision of what to do with pvObj? I have expressed my views on the matter, but if there's a general consensus to keep PVObjects in Python, then so be it. Either way, it would be nice to come to a decision so I can start reworking the analysis tools to be able to write custom pvp files.

**Comment by slundqui** [comment]
To summarize, here's what I propose to do.
- Change the output of readpvpfile to be a dictionary with 3 keys: header, time, values.
- Header will be a dictionary of key-value pairs corresponding to the header of the file.
- Time will be a 1D numpy array of size [numFrames]. 
- Values will be one of the following: 4D Numpy array for dense layers [numFrames, ny, nx, nf]; 2D Scipy coo sparse matrix for sparse layers [numFrames, ny \* nx \* nf]; 5D Numpy arrays for weights [numFrames, numArbors, nyp, nxp, nfp]
- Writepvpfile will take as arguments a dictionary with at least 2 keys: time, values. The format will be the same as the output of readpvpfile
- The header in writepvpfile will be inferred based on the size of the matrix, type of matrix, etc.
- Any additional information needed for writepvpfile (such as the ny, nx, and nf fields for the header in sparse layers) will be set as function arguments to writepvpfile.

I will start working on implementing this on Monday to provide time for people to rebut any of these proposed changes.

**Comment by slundqui** [comment]
One thing which I'm not completely certain about is the addition of the header in the output dictionary of readpvpfile. Should we stick with the Octave style, i.e.

``` python
(data, header) = readpvpfile(filename)
```

or should we go with what I proposed, i.e.

``` python
data = readpvpfile(filename)
header = data["header"]
```

**Comment by dpaiton** [comment]
I'd say 
`header = data["header"]`
is better. 

It would be nice if writepvpfile could _either_ receive fields such as nf,ny,nx as function arguments or pull them from the included header. In other words, if I want to pass writepvpfile a data object, it will first check data["header"] for the required fields and if those are missing then it will check the function arguments.

Other than that I agree with what you are proposing!

**Comment by slundqui** [comment]
Readpvpfile fixed in commit ebbe5ddc71ec35e2367020fd75ef0652b5bfb59e as proposed. Only difference between what I proposed and what is written is that I forgot the arbors dimension of weights. Namely, weights is now a 6D numpy array: [numFrames, numArbors, numKernels, nyp, nxp, nfp].

There was a comment in the file that reading pvp files from non-shared weights was not tested. I commented that code out, and it throws an error for now if you try to read a non-shared weights pvp file. Specifically, I had a question for @peteschultz. How does pvp files store shrunken patches in the case of non-shared weights?

**Comment by peteschultz** [comment]
Each data patch uses 8+datasize_(nxp_nyp*nfp) bytes in the file, whether shrunken or not, with the first 8 bytes containing the nx (16-bit), ny (16-bit), and offset (32-bit) for that patch.

The active values appear in the same part of the patch as they would if the patch were non-shrunken.  That is, for the patch in the top left corner of the layer, the active synapses are stored in the lower right corner of the patch, and so on.

**Comment by slundqui** [comment]
Quick question to @peteschultz:

I'm currently testing reading/writing pvp files to make sure what I'm writing is identical to what I'm reading. Specifically, I'm testing the sparse values read/write. Here is the header dictionary from the original pvp file I read, and the header from the written out pvpfile:

``` python
#Original:
{'ky0': 0, 'datatype': 4, 'nyprocs': 1, 'filetype': 6, 'numparams': 20, 'numrecords': 1, 'nf': 32, 'ny
Global': 240, 'nx': 240, 'ny': 240, 'nbands': 11, 'nxGlobal': 240, 'nbatch': 1, 'recordsize': 125623, 
'time': 0.0, 'kx0': 0, 'nxprocs': 1, 'datasize': 8, 'headersize': 80}
#Written:
{'ky0': 0, 'datatype': 4, 'nyprocs': 1, 'filetype': 6, 'numparams': 20, 'numrecords': 1, 'nf': 32, 'ny
Global': 240, 'nx': 240, 'ny': 240, 'nbands': 11, 'nxGlobal': 240, 'nbatch': 1, 'recordsize': 0, 'time
': 0.0, 'kx0': 0, 'nxprocs': 1, 'datasize': 8, 'headersize': 80}
```

The difference here is in the field recordsize. From the matlab version of writepvpfile, it explicitly sets record size to 0, with a comment that says the field is unnecessary. Would you happen to know what the recordsize field is used for in sparse value pvp files?

**Comment by peteschultz** [comment]
For sparse activity and sparse values files, the recordSize field is the number of active elements when the header was written, that is, the first frame.  The problem is that layer .pvp files only have a single header for the entire file, while the number of active elements changes each frame.  Hence each frame begins with the count of active neurons in that frame.  So the entry in the header is useless.  Perhaps this header entry should be written as zero by the writeSparseActivity function in fileio.cpp, just as it is for the matlab m-file.

**Comment by slundqui** [comment]
Updated writepvpfile to match readpvpfile in commit a124ab81a3ec54a4a23c4f71f652381493ba12c5. This means pytools is in a state to be used now for both reading and writing. The only thing is that non-shared weights and sparse activities (without values) are currently not supported in writing and reading, which can be added as we run into these cases being needed. Please let me know if you have any problems with it. I will be writing a wiki on these python tools soon.
